server/src/server.common.ts

```diff
@@ -181,53 +177,111 @@ export function startServer(connection: _Connection) {
 		connection.sendDiagnostics({ uri: textDocument.uri, diagnostics });
 	}
 
-	function validateYAML(textDocument: TextDocument): void {
-		connection.sendDiagnostics({
-			uri: textDocument.uri,
-			diagnostics: [],
-		});
-
+	function validateYamlSyntaxAndModel(textDocument: TextDocument): Diagnostic[] {
 		const diagnostics: Diagnostic[] = [];
 
 		const lineCounter = new LineCounter();
-		const doc = parseDocument(textDocument.getText(), {
+		const yamlDoc = parseDocument(textDocument.getText(), {
 			lineCounter,
 			keepSourceTokens: true,
-			uniqueKeys: false
 		});
 
+		const map = new YAMLSourceMap();
+		map.doMap(yamlDoc.contents);
+
 		// Basic syntax errors
-		for (const err of doc.errors) {
+		for (const err of yamlDoc.errors) {
 			diagnostics.push({ message: err.message, range: rangeFromLinePos(err.linePos) });
 		}
 
-		// Get location of model in CST
-		if (doc.has("model")) {
-			let position: {line: number, col: number};
-
-			// Get the model token and find its position
-			(doc.contents?.srcToken as BlockMap).items.forEach(i => {
-				if (i.key?.offset !== undefined && (i.key as SourceToken).source === "model") {
-					position = lineCounter.linePos(i.key?.offset);
-				}
-			});
+		// If no diagnostics, continue parsing.
+		if (diagnostics.length === 0) {
+
+			const validator = new Ajv();
```

Comment:
For performance reasons, move these out of the function. This should run only once in the lifetime of the extension, and only the validate part should run in the function